android/app/src/main/java/com/microsoft/codepush/common/managers/CodePushRestartManager.java

```diff
@@ -0,0 +1,100 @@
+package com.microsoft.codepush.common.managers;
+
+import com.microsoft.appcenter.utils.AppCenterLog;
+import com.microsoft.codepush.common.interfaces.CodePushRestartListener;
+
+import java.util.LinkedList;
+import java.util.List;
+
+import static com.microsoft.codepush.common.CodePush.LOG_TAG;
+
+/**
+ * Manager responsible for restarting the application.
+ */
+public class CodePushRestartManager {
+
+    /**
+     * Listener for restart events.
+     */
+    private CodePushRestartListener mRestartListener;
+
+    /**
+     * <code>true</code> if restart is allowed.
+     */
+    private boolean mAllowed = true;
+
+    /**
+     * <code>true</code> if application is in the process of restart.
+     */
+    private boolean mRestartInProgress = false;
+
+    /**
+     * Queue containing pending restart requests.
+     */
+    private List<Boolean> mRestartQueue = new LinkedList<>();
+
+    /**
+     * Creates an instance of {@link CodePushRestartManager}.
+     *
+     * @param codePushRestartListener listener for restart events.
+     */
+    public CodePushRestartManager(CodePushRestartListener codePushRestartListener) {
+        mRestartListener = codePushRestartListener;
+    }
+
+    /**
+     * Allows the manager to perform restarts and performs them if there are pending.
+     */
+    public void allowRestarts() {
+        AppCenterLog.logAssert(LOG_TAG, "Re-allowing restarts");
```

Comment:
Why do you use `logAssert` instead of `info` or just `log`?

Reply:
@ruslan-bikkinin Because we are using it already in other project, look at the UpdateManager, for instance

Reply:
I believe we should not use `logAssert ` method, please see https://github.com/Microsoft/AppCenter-SDK-Android/search?utf8=%E2%9C%93&q=AppCenterLog.logAssert&type=

You should use `info`, `debug` and etc. Please also investigated AppCenter-SDK-Android repo for the same situations.